<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>iVar: Manual</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">iVar
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">Manual </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><h2>Available Commands</h2>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Command  </th><th class="markdownTableHeadLeft">Description   </th></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">trim  </td><td class="markdownTableBodyLeft">Trim reads in aligned BAMs   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">variants  </td><td class="markdownTableBodyLeft">Call variants from aligned BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">filtervariants  </td><td class="markdownTableBodyLeft">Filter variants across replicates   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">consensus  </td><td class="markdownTableBodyLeft">Call consensus from aligned BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">getmasked  </td><td class="markdownTableBodyLeft">Get amplicons with primer mismatches   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">removereads  </td><td class="markdownTableBodyLeft">Remove reads from trimmed BAM file   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">version  </td><td class="markdownTableBodyLeft">Show version information   </td></tr>
<tr class="markdownTableBody" class="markdownTableRowEven">
<td class="markdownTableBodyLeft">trimadapter  </td><td class="markdownTableBodyLeft">(EXPERIMENTAL) Trim adapter sequences from reads   </td></tr>
</table>
<p>To view detailed usage for each command type <code>ivar &lt;command&gt;</code> Note : Commands maked (EXPERIMENTAL) are still under active development.</p>
<h2>Description of commands</h2>
<h4>Trim primer sequences with iVar</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar trim</div><div class="line"></div><div class="line">Usage: ivar trim -i &lt;input.bam&gt; -b &lt;primers.bed&gt; -p prefix [-m &lt;min-length&gt;] [-q &lt;min-quality&gt;] [-s &lt;sliding-window-width&gt;]</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) &lt;input.bam&gt; Indexed aligned bam file to trim primers and quality</div><div class="line">           -b    (Required) &lt;primers.bed&gt; BED file with primer sequences and positions</div><div class="line">           -m    Minimum length of read to retain after trimming (Default: 30)</div><div class="line">           -q    Minimum quality threshold for sliding window to pass (Default: 20)</div><div class="line">           -s    Width of sliding window (Default: 4)</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix for the output BAM file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --><h4>Call variants with iVar</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar variants</div><div class="line"></div><div class="line">Usage: samtools mpileup -A -d 300000 --reference &lt;reference-fasta&gt; -Q 0 -F 0 &lt;input-bam&gt; | ivar variants -p prefix [-q &lt;min-quality&gt;] [-t &lt;min-frequency-threshold&gt;]</div><div class="line"></div><div class="line">Note : samtools mpileup output must be piped into ivar variants</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -q    &lt;min-quality&gt; Minimum quality threshold to count base (Default: 20)</div><div class="line">           -t    &lt;min-frequency-threshold&gt; Minimum frequency threshold(0 - 1) to call variants (Default: 0.03)</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix, prefix for the output tsv variant file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --><h4>Filter variants across replicates with iVar</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar filtervariants</div><div class="line">Usage: ivar filtervariants [-p prefix] replicate-one.tsv replicate-two.csv ...</div><div class="line"></div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix, prefix for the filtered tsv file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --><h4>Call a consensus sequences from an aligned BAM file.</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar consensus</div><div class="line">Usage: samtools mpileup -A -d 300000 -Q 0 -F 0 [&lt;input-bam&gt;] | ivar consensus [-p prefix]</div><div class="line"></div><div class="line">Note : samtools mpileup output must be piped into &quot;ivar consensus&quot;</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -q    Minimum quality threshold to count base (Default: 20)</div><div class="line">           -t    Threshold(0 - 100) to call consensus (Default: 0)</div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix, prefix for the output tsv file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --><h4>Get primers with mismatches to the reference sequence.</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar getmasked</div><div class="line">Usage: ivar getmasked [-i &lt;input-filtered-tsv&gt;] [-b &lt;bed-file&gt;]</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) Input filtered variants tsv generated from &quot;ivar filtervariants&quot;</div><div class="line">           -b    (Required) Bed file with primer indices</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --><h4>Remove reads associated with mismatched primer indices</h4>
<p>Command: </p><div class="fragment"><div class="line">ivar removereads</div><div class="line">Usage: ivar removereads [-i &lt;input-bam&gt;] [-p prefix] primer-index-1 primer-index-2 primer-index-3 primer-index-4 ...</div><div class="line"></div><div class="line">Input Options    Description</div><div class="line">           -i    (Required) Input BAM file run through `ivar trim` command whcih adds the primer number to BAM auxillary data</div><div class="line">Output Options   Description</div><div class="line">           -p    (Required) Prefix, prefix for the filtered BAM file</div></div><!-- fragment --><p>Example Usage: </p><div class="fragment"></div><!-- fragment --> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.14
</small></address>
</body>
</html>
