\hypertarget{manualpage_autotoc_md0}{}\section{Available Commands}\label{manualpage_autotoc_md0}
\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{2}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Command  }&\textbf{ Description   }\\\cline{1-2}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Command  }&\textbf{ Description   }\\\cline{1-2}
\endhead
trim  &Trim reads in aligned B\+AM   \\\cline{1-2}
variants  &Call variants from aligned B\+AM file   \\\cline{1-2}
filtervariants  &Filter variants across replicates   \\\cline{1-2}
consensus  &Call consensus from aligned B\+AM file   \\\cline{1-2}
getmasked  &Get amplicons with primer mismatches   \\\cline{1-2}
removereads  &Remove reads from trimmed B\+AM file   \\\cline{1-2}
version  &Show version information   \\\cline{1-2}
trimadapter  &(E\+X\+P\+E\+R\+I\+M\+E\+N\+T\+AL) Trim adapter sequences from reads   \\\cline{1-2}
\end{longtabu}


To view detailed usage for each command type {\ttfamily ivar $<$command$>$} Note \+: Commands maked (E\+X\+P\+E\+R\+I\+M\+E\+N\+T\+AL) are still under active development.\hypertarget{manualpage_autotoc_md1}{}\section{Description of commands}\label{manualpage_autotoc_md1}
\hypertarget{manualpage_autotoc_md2}{}\subsection{Trim primer sequences with i\+Var}\label{manualpage_autotoc_md2}
i\+Var uses primer positions supplied in a B\+ED file to soft clip primer sequences from an aligned, sorted and indexed B\+AM file. Following this, the reads are trimmed based on a quality threshold(\+Default\+: 20). To do the quality trimming, i\+Var uses a sliding window approach(\+Default\+: 4). The windows slides from the 5\textquotesingle{} end to the 3\textquotesingle{} end and if at any point the average base quality in the window falls below the threshold, the remaining read is soft clipped. If after trimming, the length of the read is greater than the minimum length specified(\+Default\+: 30), the read is written to the new trimmed B\+AM file.

To sort and index an aligned B\+AM file, the following command can be used,


\begin{DoxyCode}
# Input file - test.bam
samtools sort -o test.sorted.bam test.bam && samtools index test.sorted.bam.
\end{DoxyCode}


{\bfseries Note}\+: All the trimming in i\+Var is done by soft-\/clipping reads in an aligned B\+AM file. This information is lost if reads are extracted in fastq or fasta format from the trimmed B\+AM file.

Command\+: 
\begin{DoxyCode}
ivar trim

Usage: ivar trim -i <input.bam> -b <primers.bed> -p <prefix> [-m <min-length>] [-q <min-quality>] [-s
       <sliding-window-width>]

Input Options    Description
           -i    (Required) Indexed aligned bam file to trim primers and quality
           -b    (Required) BED file with primer sequences and positions
           -m    Minimum length of read to retain after trimming (Default: 30)
           -q    Minimum quality threshold for sliding window to pass (Default: 20)
           -s    Width of sliding window (Default: 4)

Output Options   Description
           -p    (Required) Prefix for the output BAM file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
ivar trim -b test\_primers.bed -p test.trimmed -i test.bam -q 15 -m 50 -s 4
\end{DoxyCode}


The command above will produce a trimmed B\+AM file test.\+trimmed.\+bam after trimming the aligned reads in test.\+bam using the primer positions specified in test\+\_\+primers.\+bed and a minimum quality threshold of {\bfseries 15}, minimum read length of {\bfseries 50} and a sliding window of {\bfseries 4}.

Example B\+ED file


\begin{DoxyCode}
Puerto  28  52  400\_1\_out\_L 60  +
Puerto  482 504 400\_1\_out\_R 60  -
Puerto  359 381 400\_2\_out\_L 60  +
Puerto  796 818 400\_2\_out\_R 60  -
Puerto  658 680 400\_3\_out\_L*    60  +
Puerto  1054    1076    400\_3\_out\_R*    60  -
.
.
.
.
\end{DoxyCode}
\hypertarget{manualpage_autotoc_md3}{}\subsection{Call variants with i\+Var}\label{manualpage_autotoc_md3}
i\+Var uses the output of the {\ttfamily samtools mpileup} command to call variants -\/ single nucleotide variants(\+S\+N\+Vs) and indels. In order to call variants correctly, the {\ttfamily samtools mpileup} command must be run with the reference fasta supplied using the options(--reference or -\/f). The output of {\ttfamily samtools pileup} is then piped into {\ttfamily ivar variants} to generate a .tsv file with the following fields,

There are two parameters that can be set for variant calling using i\+Var -\/ minimum quality(\+Default\+: 20) and minimum frequency(Default\+: 0.\+03). Minimum quality is the minimum quality for a base to be used in frequency calculations at a given position. Minimum frequency is the minimum frequency required for a S\+NV or indel to be reported.

Command\+: 
\begin{DoxyCode}
ivar variants

Usage: samtools mpileup -A -d 300000 --reference <reference-fasta> -B -Q 0 -F 0 <input.bam> | ivar variants
       -p <prefix> [-q <min-quality>] [-t <min-frequency-threshold>]

Note : samtools mpileup output must be piped into ivar variants

Input Options    Description
           -q    Minimum quality score threshold to count base (Default: 20)
           -t    Minimum frequency threshold(0 - 1) to call variants (Default: 0.03)

Output Options   Description
           -p    (Required) Prefix for the output tsv variant file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
samtools mpileup --reference test\_reference.fa -A -d 600000 -F 0 -B -Q 0 test.trimmed.bam | ivar variants
       -p test -q 20 -t 0.03
\end{DoxyCode}


The command above will generate a test.\+tsv file.

Example of output .tsv file.


\begin{DoxyCode}
REGION  POS REF ALT REF\_DP  REF\_RV  REF\_QUAL    ALT\_DP  ALT\_RV  ALT\_QUAL    ALT\_FREQ    TOTAL\_DP    PVAL   
       PASS
Puerto  77  A   G   9204    0   37  2843    0   37  0.235973    12048   0   TRUE
Puerto  110 G   A   11289   0   37  443 0   37  0.0377568   11733   1.44315e-117    TRUE
Puerto  118 C   +AA 11613   0   37  365 0   20  0.0314087   11621   2.94565e-30 TRUE
Puerto  118 C   +A  11613   0   37  635 0   20  0.0546425   11621   2.95259e-84 TRUE
\end{DoxyCode}


Description

\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{2}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Field  }&\textbf{ Description   }\\\cline{1-2}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Field  }&\textbf{ Description   }\\\cline{1-2}
\endhead
R\+E\+G\+I\+ON  &Region from B\+AM file   \\\cline{1-2}
P\+OS  &Position on reference sequence   \\\cline{1-2}
R\+EF  &Reference base   \\\cline{1-2}
A\+LT  &Alternate Base   \\\cline{1-2}
R\+E\+F\+\_\+\+DP  &Depth of reference base   \\\cline{1-2}
R\+E\+F\+\_\+\+RV  &Depth of reference base on reverse reads   \\\cline{1-2}
R\+E\+F\+\_\+\+Q\+U\+AL  &Mean quality of reference base   \\\cline{1-2}
A\+L\+T\+\_\+\+DP  &Depth of alternate base   \\\cline{1-2}
A\+L\+T\+\_\+\+RV  &Deapth of alternate base on reverse reads   \\\cline{1-2}
A\+L\+T\+\_\+\+Q\+U\+AL  &Mean quality of alternate base   \\\cline{1-2}
A\+L\+T\+\_\+\+F\+R\+EQ  &Frequency of alternate base   \\\cline{1-2}
T\+O\+T\+A\+L\+\_\+\+DP  &Total depth at position   \\\cline{1-2}
P\+V\+AL  &p-\/value of fisher\textquotesingle{}s exact test   \\\cline{1-2}
P\+A\+SS  &Result of p-\/value $<$= 0.\+05   \\\cline{1-2}
\end{longtabu}


{\bfseries Note}\+: Please use the -\/B options with {\ttfamily samtools mpileup} to call variants and generate consensus. When a reference sequence is supplied, the quality of the reference base is reduced to 0 (A\+S\+C\+II\+: !) in the mpileup output. Disabling B\+AQ with -\/B seems to fix this. This was tested in samtools 1.\+7 and 1.\+8.\hypertarget{manualpage_autotoc_md4}{}\subsection{Filter variants across replicates with i\+Var}\label{manualpage_autotoc_md4}
Under the hood, i\+Var calls an Awk script to get an intersection of variants(in .tsv files) called from any number of replicates. This intersection will filter out any S\+N\+Vs that do not pass the filters(in the variant calling step) in all the replicates. Fields that are different across replicates(fields apart from R\+E\+G\+I\+O\+N, P\+O\+S, R\+E\+F, A\+L\+T) will have the filename added as a suffix.

Command\+: 
\begin{DoxyCode}
ivar filtervariants

Usage: ivar filtervariants -p <prefix> replicate-one.tsv replicate-two.tsv ...

Input: Variant tsv files for each replicate

Output Options   Description
           -p    (Required) Prefix for the output filtered tsv file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
ivar filtervariants -p test.filtered test\_rep1.tsv test\_rep2.tsv test\_rep3.tsv
\end{DoxyCode}


The command above will prodoce an output .tsv file test.\+filtered.\+tsv.

Example output of filtered .tsv file from two files test\+\_\+rep1.\+tsv and test\+\_\+rep2.\+tsv


\begin{DoxyCode}
REGION  POS REF ALT REF\_DP\_test\_rep2.tsv    REF\_RV\_test\_rep2.tsv    REF\_QUAL\_test\_rep2.tsv 
       ALT\_DP\_test\_rep2.tsv    ALT\_RV\_test\_rep2.tsv    ALT\_QUAL\_test\_rep2.tsv  ALT\_FREQ\_test\_rep2.tsv  TOTAL\_DP\_test\_rep2.tsv 
       PVAL\_test\_rep2.tsv  PASS\_test\_rep2.tsv  REF\_DP\_test\_rep1.tsv    REF\_RV\_test\_rep1.tsv    REF\_QUAL\_test\_rep1.tsv 
       ALT\_DP\_test\_rep1.tsv    ALT\_RV\_test\_rep1.tsv    ALT\_QUAL\_test\_rep1.tsv  ALT\_FREQ\_test\_rep1.tsv 
       TOTAL\_DP\_test\_rep1.tsv  PVAL\_test\_rep1.tsv  PASS\_test\_rep1.tsv  
Puerto  77  A   G   9204    0   37  2843    0   37  0.235973    12048   0   TRUE    9204    0   37  2843   
       0   37  0.235973    12048   0   TRUE    
Puerto  110 G   A   11289   0   37  443 0   37  0.0377568   11733   1.44315e-117    TRUE    11289   0   37 
       443 0   37  0.0377568   11733   1.44315e-117    TRUE    
Puerto  118 C   +AA 11613   0   37  365 0   20  0.0314087   11621   2.94565e-30 TRUE    11613   0   37  365
       0   20  0.0314087   11621   2.94565e-30 TRUE    
Puerto  118 C   +A  11613   0   37  635 0   20  0.0546425   11621   2.95259e-84 TRUE    11613   0   37  635
       0   20  0.0546425   11621   
\end{DoxyCode}


Description of fields

\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{3}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ No  }&\textbf{ Field  }&\textbf{ Description   }\\\cline{1-3}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ No  }&\textbf{ Field  }&\textbf{ Description   }\\\cline{1-3}
\endhead
1  &R\+E\+G\+I\+ON  &Common region across all replicate variant tsv files   \\\cline{1-3}
2  &P\+OS  &Common position across all variant tsv files   \\\cline{1-3}
3  &R\+EF  &Common reference base across all variant tsv files   \\\cline{1-3}
4  &A\+LT  &Common alternate base across all variant tsv files   \\\cline{1-3}
5  &R\+E\+F\+\_\+\+D\+P\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Depth of reference base in replicate 1   \\\cline{1-3}
6  &R\+E\+F\+\_\+\+R\+V\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Depth of reference base on reverse reads in replicate 1   \\\cline{1-3}
7  &R\+E\+F\+\_\+\+Q\+U\+A\+L\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Mean quality of reference base in replicate 1   \\\cline{1-3}
8  &A\+L\+T\+\_\+\+D\+P\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Depth of alternate base in replicate 1   \\\cline{1-3}
9  &A\+L\+T\+\_\+\+R\+V\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Deapth of alternate base on reverse reads in replicate 1   \\\cline{1-3}
10  &A\+L\+T\+\_\+\+Q\+U\+A\+L\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Mean quality of alternate base in replicate 1   \\\cline{1-3}
11  &A\+L\+T\+\_\+\+F\+R\+E\+Q\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Frequency of alternate base in replicate 1   \\\cline{1-3}
12  &T\+O\+T\+A\+L\+\_\+\+D\+P\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Total depth at position in replicate 1   \\\cline{1-3}
13  &P\+V\+A\+L\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &p-\/value of fisher\textquotesingle{}s exact test in replicate 1   \\\cline{1-3}
14  &P\+A\+S\+S\+\_\+$<$rep1-\/tsv-\/file-\/name$>$  &Result of p-\/value $<$= 0.\+05 in replicate 1   \\\cline{1-3}
15  &Continue rows 5 -\/ 14 for every replicate provided  &\\\cline{1-3}
\end{longtabu}
\hypertarget{manualpage_autotoc_md5}{}\subsection{Generate a consensus sequences from an aligned B\+A\+M file}\label{manualpage_autotoc_md5}
To generate a consensus sequence i\+Var uses the output of {\ttfamily samtools mpileup} command. The mpileup output must be piped into {\ttfamily ivar consensus}. There are five parameters that can be set -\/ minimum quality(\+Default\+: 20), minimum frequency threshold(\+Default\+: 0), minimum depth to call a consensus(\+Default\+: 1), a flag to exclude nucleotides from regions with depth less than the minimum depth and a character to call in regions with coverage lower than the speicifed minimum depth(Default\+: \textquotesingle{}-\/\textquotesingle{}). Minimum quality is the minimum quality of a base to be considered in calculations of variant frequencies at a given position. Minimum frequency threshold is the minimum frequency that a base must match to be called as the consensus base at a position. If one base is not enough to match a given frequency, then an ambigious nucleotide is called at that position. Minimum depth is the minimum required depth to call a consensus. If \textquotesingle{}-\/k\textquotesingle{} flag is set then these regions are not included in the consensus sequence. If \textquotesingle{}-\/k\textquotesingle{} is not set then by default, a \textquotesingle{}-\/\textquotesingle{} is called in these regions. You can also specfy which character you want to add to the consensus to cover regions with depth less than the minimum depth. This can be done using -\/n options. It takes onr of two values\+: \textquotesingle{}-\/\textquotesingle{} or \textquotesingle{}N\textquotesingle{}.

As an example, consider a position with 6\+As, 3\+Ts and 1C. The table below shows the consensus nucleotide called at different frequencies.

\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{2}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Minimum frequency threshold  }&\textbf{ Consensus   }\\\cline{1-2}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Minimum frequency threshold  }&\textbf{ Consensus   }\\\cline{1-2}
\endhead
0  &A   \\\cline{1-2}
0.\+5  &A   \\\cline{1-2}
0.\+6  &A   \\\cline{1-2}
0.\+7  &W(\+A or T)   \\\cline{1-2}
0.\+9  &W (A or T)   \\\cline{1-2}
1  &H (A or T or C)   \\\cline{1-2}
\end{longtabu}


If there are two nucleotides at the same frequency, both nucleotides are used to call an ambigious base as the consensus. As an example, consider a position wiht 6 Ts, 2\+As and 2 Gs. The table below shows the consensus nucleotide called at different frequencies.

\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{2}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Minimum frequency threshold  }&\textbf{ Consensus   }\\\cline{1-2}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ Minimum frequency threshold  }&\textbf{ Consensus   }\\\cline{1-2}
\endhead
0  &T   \\\cline{1-2}
0.\+5  &T   \\\cline{1-2}
0.\+6  &T   \\\cline{1-2}
0.\+7  &D(\+A or T or G)   \\\cline{1-2}
0.\+9  &D(\+A or T or G)   \\\cline{1-2}
1  &D(\+A or T or G)   \\\cline{1-2}
\end{longtabu}


The output of the command is a fasta file with the consensus sequence and a .txt file with the average quality of every base used to generate the consensus at each position. {\itshape For insertions, the quality is set to be the minimum quality threshold since mpileup doesn\textquotesingle{}t give the quality of bases in insertions.}

Command\+: 
\begin{DoxyCode}
ivar consensus

Usage: samtools mpileup -A -d 300000 -Q 0 -F 0 <input.bam> | ivar consensus -p <prefix>

Note : samtools mpileup output must be piped into ivar consensus

Input Options    Description
           -q    Minimum quality score threshold to count base (Default: 20)
           -t    Minimum frequency threshold(0 - 1) to call consensus. (Default: 0)
                 Frequently used thresholds | Description
                 ---------------------------|------------
                                          0 | Majority or most common base
                                        0.2 | Bases that make up atleast 20% of the depth at a position
                                        0.5 | Strict or bases that make up atleast 50% of the depth at a
       position
                                        0.9 | Strict or bases that make up atleast 90% of the depth at a
       position
                                          1 | Identical or bases that make up 100% of the depth at a
       position. Will have highest ambiguities
           -m    Minimum depth to call consensus(Default: 1)
           -k    If '-k' flag is added, regions with depth less than minimum depth will not be added to the
       consensus sequence. Using '-k' will override any option specified using -n
           -n    (N/-) Character to print in regions with less than minimum coverage(Default: -)

Output Options   Description
           -p    (Required) Prefix for the output fasta file and quality file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
samtools mpileup -d 1000 -A -Q 0 -F 0 test.bam | ivar consensus -p test -q 20 -t 0
\end{DoxyCode}


The command above will produce a test.\+fa fasta file with the consensus sequence and a test.\+qual.\+txt with the average quality of each base in the consensus sequence.\hypertarget{manualpage_autotoc_md6}{}\subsection{Get primers with mismatches to the reference sequence}\label{manualpage_autotoc_md6}
i\+Var uses a .tsv file with variants to get the zero based indices(based on the B\+E\+D file) of mismatched primers. The output is the primer indices delimited by a space. The output is written to stdout and can be written to a file by redirecting output into a file using {\ttfamily $>$}.

Command\+: 
\begin{DoxyCode}
ivar getmasked
Usage: ivar getmasked -i <input-filtered.tsv> -b <primers.bed> -p <prefix>
Note: This step is used only for amplicon-based sequencing.

Input Options    Description
           -i    (Required) Input filtered variants tsv generated from ivar filtervariants
           -b    (Required) BED file with primer sequences and positions

Output Options   Description
           -p    (Required) Prefix for the output text file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
ivar getmasked -i test.filtered.tsv -b primers.bed -p test.masked.txt
\end{DoxyCode}


The command above produces an output file -\/ test.\+masked.\+txt.

Example Output\+:


\begin{DoxyCode}
1 2 7 8
\end{DoxyCode}
\hypertarget{manualpage_autotoc_md7}{}\subsection{Remove reads associated with mismatched primer indices}\label{manualpage_autotoc_md7}
This command accepts an aligned, sorted and indexed B\+AM file trimmed using {\ttfamily ivar trim} and removes the reads corresponding to the supplied primer indices, which is the output of {\ttfamily ivar getmasked} command. Under the hood, {\ttfamily ivar trim} adds the zero based primer index(based on the B\+E\+D file) to the B\+AM auxillary data for every read. Hence, ivar removereads will only work on B\+AM files that have been trimmed using {\ttfamily ivar trim}.

Command\+: 
\begin{DoxyCode}
ivar removereads

Usage: ivar removereads -i <input.trimmed.bam> -p <prefix> -t <text-file-with-primer-indices>
Note: This step is used only for amplicon-based sequencing.

Input Options    Description
           -i    (Required) Input BAM file  trimmed with ivar trim. Must be sorted and indexed, which can
       be done using sort\_index\_bam.sh
           -t    (Required) Text file with primer indices separated by spaces. This is the output of
       getmasked command.

Output Options   Description
           -p    (Required) Prefix for the output filtered BAM file
\end{DoxyCode}


Example Usage\+: 
\begin{DoxyCode}
ivar trim -i test.bam -p test.trimmed
ivar removereads -i test.trimmed.bam -p test.trimmed.masked.bam -t test.masked.txt
\end{DoxyCode}


The {\ttfamily ivar trim} command above trims test.\+bam and produced test.\+trimmed.\+bam with the primer indice data added. The {\ttfamily ivar removereads} command produces an output file -\/ test.\+trimmed.\+masked.\+bam after removing all the reads corresponding to primer indices -\/ 1,2,7 and 8.\hypertarget{manualpage_autotoc_md8}{}\subsection{(\+Experimental) trimadapter}\label{manualpage_autotoc_md8}
{\bfseries Note\+: This feature is under active development and not completely validated yet.}

trimadapter in i\+Var can be used to trim adapter sequences from fastq files using a supplied fasta file. 